In [1]:

# IMPORTS
import numpy as np
import pandas as pd

#Fin Data Sources
import yfinance as yf
import pandas_datareader as pdr

#Data viz
import plotly.graph_objs as go
import plotly.express as px

import time
from datetime import date

# for graphs
import matplotlib.pyplot as plt

### Question 1: IPO Filings Web Scraping and Data Processing

**What's the total sum ($m) of 2023 filings that happened on Fridays?**

Re-use the [Code Snippet 1] example to get the data from web for this endpoint: https://stockanalysis.com/ipos/filings/
Convert the 'Filing Date' to datetime(), 'Shares Offered' to float64 (if '-' is encountered, populate with NaNs).
Define a new field 'Avg_price' based on the "Price Range", which equals to NaN if no price is specified, to the price (if only one number is provided), or to the average of 2 prices (if a range is given).
You may be inspired by the function `extract_numbers()` in [Code Snippet 4], or you can write your own function to "parse" a string.
Define a column "Shares_offered_value", which equals to "Shares Offered" * "Avg_price" (when both columns are defined; otherwise, it's NaN)

Find the total sum in $m (millions of USD, closest INTEGER number) for all filings during 2023, which happened on Fridays (`Date.dt.dayofweek()==4`). You should see 32 records in total, 25 of it is not null.

(additional: you can read about [S-1 IPO filing](https://www.dfinsolutions.com/knowledge-hub/thought-leadership/knowledge-resources/what-s-1-ipo-filing) to understand the context)

In [8]:
import pandas as pd
import requests

headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3',
}

url = "https://stockanalysis.com/ipos/filings/"
response = requests.get(url, headers=headers)

ipo_dfs = pd.read_html(response.text)

C:\Users\kiramishima\AppData\Local\Temp\ipykernel_25652\2819786403.py:11: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  ipo_dfs = pd.read_html(response.text)


In [9]:
ipos_2023 = ipo_dfs[0]
ipos_2023.head()

Filing Date Symbol                           Company Name Price Range  \
0   May 3, 2024    TBN         Tamboran Resources Corporation           -   
1  Apr 29, 2024   HWEC                   HW Electro Co., Ltd.       $3.00   
2  Apr 29, 2024   DTSQ  DT Cloud Star Acquisition Corporation      $10.00   
3  Apr 26, 2024   EURK                Eureka Acquisition Corp      $10.00   
4  Apr 26, 2024    HDL    Super Hi International Holding Ltd.           -   

  Shares Offered  
0              -  
1        3750000  
2        6000000  
3        5000000  
4              -

In [10]:
ipos_2023['Filing Date'] = pd.to_datetime(ipos_2023['Filing Date'], format='%b %d, %Y')

In [17]:
ipos_2023['Shares Offered'].replace({'-': np.nan}, inplace=True)
ipos_2023['Shares Offered'] = ipos_2023['Shares Offered'].astype(np.float64)

In [19]:
import re

# Code Snippet 4
def extract_numbers(input_string):
    split_string = input_string.split(" - ")
    y_match = re.search(r'(\d+.\d)', split_string[0])
    if len(split_string) > 1:
      m_match = re.search(r'(\d+.\d)', split_string[1])
      y1_number, m1_number = float(y_match.group(1)) if y_match else 0, float(m_match.group(1)) if m_match else 0
      return (y1_number + m1_number)/len(split_string)
    else:
      y0 = float(y_match.group(1)) if y_match else 0
      return y0

In [20]:
ipos_2023['Avg_price'] = ipos_2023['Price Range'].apply(lambda x: extract_numbers(x))

In [22]:
ipos_2023['Shares_offered_value'] = ipos_2023['Shares Offered'] * ipos_2023['Avg_price']

In [23]:
ipos_2023.head()

Filing Date Symbol                           Company Name Price Range  \
0  2024-05-03    TBN         Tamboran Resources Corporation           -   
1  2024-04-29   HWEC                   HW Electro Co., Ltd.       $3.00   
2  2024-04-29   DTSQ  DT Cloud Star Acquisition Corporation      $10.00   
3  2024-04-26   EURK                Eureka Acquisition Corp      $10.00   
4  2024-04-26    HDL    Super Hi International Holding Ltd.           -   

   Shares Offered  Avg_price  Shares_offered_value  
0             NaN        0.0                   NaN  
1       3750000.0        3.0            11250000.0  
2       6000000.0       10.0            60000000.0  
3       5000000.0       10.0            50000000.0  
4             NaN        0.0                   NaN

In [28]:
ipos_filter = ipos_2023[(ipos_2023['Filing Date'].dt.dayofweek == 4)  & (ipos_2023['Filing Date'].dt.year == 2023)]

In [30]:
ipos_filter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32 entries, 50 to 165
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Filing Date           32 non-null     datetime64[ns]
 1   Symbol                32 non-null     object        
 2   Company Name          32 non-null     object        
 3   Price Range           32 non-null     object        
 4   Shares Offered        25 non-null     float64       
 5   Avg_price             32 non-null     float64       
 6   Shares_offered_value  25 non-null     float64       
dtypes: datetime64[ns](1), float64(3), object(3)
memory usage: 2.0+ KB


In [31]:
total = ipos_filter['Shares_offered_value'].sum()
f"Total sum in millions: ${round(total/1000000)}M"

'total sum in millions: $286M'

### Question 2:  IPOs "Fixed days hold" strategy


**Find the optimal number of days X (between 1 and 30), where 75% quantile growth is the highest?**


Reuse [Code Snippet 1] to retrieve the list of IPOs from 2023 and 2024 (from URLs: https://stockanalysis.com/ipos/2023/ and https://stockanalysis.com/ipos/2024/). 
Get all OHLCV daily prices for all stocks with an "IPO date" before March 1, 2024 ("< 2024-03-01") - 184 tickers (without 'RYZB'). Please remove 'RYZB', as it is no longer available on Yahoo Finance. 

Sometimes you may need to adjust the symbol name (e.g., 'IBAC' on stockanalysis.com -> 'IBACU' on Yahoo Finance) to locate OHLCV prices for all stocks. Also, you can see the ticker changes using this [link](https://stockanalysis.com/actions/changes/).
Some of the tickers (like 'DYCQ' and 'LEGT') were on the market less than 30 days (11 and 21 days, respectively). Let's leave them in the dataset; it just means that you couldn't hold them for more days than they were listed.

Let's assume you managed to buy a new stock (listed on IPO) on the first day at the [Adj Close] price]. Your strategy is to hold for exactly X full days (where X is between 1 and 30) and sell at the "Adj. Close" price in X days (e.g., if X=1, you sell on the next day).
Find X, when the 75% quantile growth (among 185 investments) is the highest. 

HINTs:
* You can generate 30 additional columns: growth_future_1d ... growth_future_30d, join that with the table of min_dates (first day when each stock has data on Yahoo Finance), and perform vector operations on the resulting dataset.
* You can use the `DataFrame.describe()` function to get mean, min, max, 25-50-75% quantiles.


Additional: 
* You can also ensure that the mean and 50th percentile (median) investment returns are negative for most X values, implying a wager for a "lucky" investor who might be in the top 25%.
* What's your recommendation: Do you suggest pursuing this strategy for an optimal X?

In [45]:
from io import StringIO
import pandas as pd
import requests

def get_ipo_data(year):
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3',
    }
    url = f'https://stockanalysis.com/ipos/{year}/'
    print(f'Getting - {url}')
    response = requests.get(url, headers=headers)
    html_io = StringIO(response.text)
    print('Finishing')
    return pd.read_html(html_io)

In [46]:
# 2023
ipo_dfs = get_ipo_data(2023)

ipos_2023 = ipo_dfs[0]
ipos_2023.info()

Getting - https://stockanalysis.com/ipos/2023/
Finishing
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154 entries, 0 to 153
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   IPO Date      154 non-null    object
 1   Symbol        154 non-null    object
 2   Company Name  154 non-null    object
 3   IPO Price     154 non-null    object
 4   Current       154 non-null    object
 5   Return        154 non-null    object
dtypes: object(6)
memory usage: 7.3+ KB


In [48]:
# 2024
ipo_dfs = get_ipo_data(2024)

ipos_2024 = ipo_dfs[0]
ipos_2024.info()

Getting - https://stockanalysis.com/ipos/2024/
Finishing
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   IPO Date      64 non-null     object
 1   Symbol        64 non-null     object
 2   Company Name  64 non-null     object
 3   IPO Price     64 non-null     object
 4   Current       64 non-null     object
 5   Return        64 non-null     object
dtypes: object(6)
memory usage: 3.1+ KB


In [50]:
df_collect = []
for year in range(2023, 2025):
    ipo_dfs = get_ipo_data(year)

    df_collect.append(ipo_dfs[0])

df_ipos = pd.concat(df_collect, ignore_index=True)

Getting - https://stockanalysis.com/ipos/2023/
Finishing
Getting - https://stockanalysis.com/ipos/2024/
Finishing


In [51]:
df_ipos['IPO Date'] = pd.to_datetime(df_ipos['IPO Date'], format="%b %d, %Y")

In [52]:
df_ipos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 218 entries, 0 to 217
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   IPO Date      218 non-null    datetime64[ns]
 1   Symbol        218 non-null    object        
 2   Company Name  218 non-null    object        
 3   IPO Price     218 non-null    object        
 4   Current       218 non-null    object        
 5   Return        218 non-null    object        
dtypes: datetime64[ns](1), object(5)
memory usage: 10.3+ KB


In [53]:
df_ipos.head()

IPO Date Symbol                   Company Name IPO Price Current   Return
0 2023-12-27   IROH  Iron Horse Acquisitions Corp.    $10.00  $10.05    0.50%
1 2023-12-19   LGCB             Linkage Global Inc     $4.00   $2.91  -27.25%
2 2023-12-15    ZKH              ZKH Group Limited    $15.50  $12.95  -19.97%
3 2023-12-15   BAYA       Bayview Acquisition Corp    $10.00  $10.18    1.80%
4 2023-12-14   INHD             Inno Holdings Inc.     $4.00   $0.62  -84.45%

In [54]:
df_filtered_ipos = df_ipos[(df_ipos['IPO Date'] < '2024-03-01') & (df_ipos['Symbol'] != 'RYZB')]

In [55]:
df_filtered_ipos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 184 entries, 0 to 217
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   IPO Date      184 non-null    datetime64[ns]
 1   Symbol        184 non-null    object        
 2   Company Name  184 non-null    object        
 3   IPO Price     184 non-null    object        
 4   Current       184 non-null    object        
 5   Return        184 non-null    object        
dtypes: datetime64[ns](1), object(5)
memory usage: 10.1+ KB


In [57]:
lista = []
for idx, val in df_filtered_ipos['Symbol'].items():
    if val in ['IBAC', 'PTHR', 'BKHA']:
        lista.append(val+'U')
    elif val == 'SBXC':
        lista.append(val+'-UN')
    else:
        lista.append(val)

In [58]:
lista

['IROH',
 'LGCB',
 'ZKH',
 'BAYA',
 'INHD',
 'AFJK',
 'GSIW',
 'FEBO',
 'CLBR',
 'ELAB',
 'RR',
 'DDC',
 'SHIM',
 'GLAC',
 'SGN',
 'HG',
 'CRGX',
 'ANSC',
 'AITR',
 'GVH',
 'LXEO',
 'PAPL',
 'ATGL',
 'MNR',
 'WBUY',
 'NCL',
 'BIRK',
 'GMM',
 'PMEC',
 'LRHC',
 'GPAK',
 'SPKL',
 'QETA',
 'MSS',
 'ANL',
 'SYRA',
 'VSME',
 'LRE',
 'TURB',
 'MDBH',
 'KVYO',
 'CART',
 'DTCK',
 'NMRA',
 'ARM',
 'SPPL',
 'NWGL',
 'SWIN',
 'IVP',
 'NNAG',
 'SRM',
 'SPGC',
 'LQR',
 'NRXS',
 'FTEL',
 'MIRA',
 'PXDT',
 'HRYU',
 'CTNT',
 'SRFM',
 'PRZO',
 'HYAC',
 'KVAC',
 'JNVR',
 'ELWS',
 'WRNT',
 'TSBX',
 'ODD',
 'APGE',
 'NETD',
 'SGMT',
 'BOWN',
 'SXTP',
 'PWM',
 'VTMX',
 'INTS',
 'SVV',
 'KGS',
 'FIHL',
 'GENK',
 'BUJA',
 'BOF',
 'AZTR',
 'CAVA',
 'ESHA',
 'ATMU',
 'ATS',
 'IPXX',
 'CWD',
 'SGE',
 'SLRN',
 'ALCY',
 'KVUE',
 'GODN',
 'TRNR',
 'AACT',
 'JYD',
 'USGO',
 'UCAR',
 'WLGS',
 'TPET',
 'TCJH',
 'GDTC',
 'VCIG',
 'GDHG',
 'ARBB',
 'ISPR',
 'MGIH',
 'MWG',
 'HSHP',
 'SFWL',
 'SYT',
 'HKIT',
 'CHSN',
 'T

In [59]:
ticker_df = []

for ticker in lista:
  df_ticker = yf.download(tickers = ticker,
                        period = "max",
                        interval = "1d")
  df_ticker['Symbol'] = ticker
  df_ticker["adj_close_minus_1"] = df_ticker['Adj Close'].shift(-1)
  df_ticker["adj_close_growth_1d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_1"] - 1
  #df_ticker = df_ticker.drop("adj_close_minus_1", axis=1)
  df_ticker["adj_close_minus_2"] = df_ticker['Adj Close'].shift(-2)
  df_ticker["adj_close_growth_2d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_2"] - 1
  df_ticker["adj_close_minus_3"] = df_ticker['Adj Close'].shift(-3)
  df_ticker["adj_close_growth_3d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_3"] - 1
  df_ticker["adj_close_minus_4"] = df_ticker['Adj Close'].shift(-4)
  df_ticker["adj_close_growth_4d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_4"] - 1
  df_ticker["adj_close_minus_5"] = df_ticker['Adj Close'].shift(-5)
  df_ticker["adj_close_growth_5d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_5"] - 1
  df_ticker["adj_close_minus_6"] = df_ticker['Adj Close'].shift(-6)
  df_ticker["adj_close_growth_6d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_6"] - 1
  df_ticker["adj_close_minus_7"] = df_ticker['Adj Close'].shift(-7)
  df_ticker["adj_close_growth_7d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_7"] - 1
  df_ticker["adj_close_minus_8"] = df_ticker['Adj Close'].shift(-8)
  df_ticker["adj_close_growth_8d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_8"] - 1
  df_ticker["adj_close_minus_9"] = df_ticker['Adj Close'].shift(-9)
  df_ticker["adj_close_growth_9d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_9"] - 1
  df_ticker["adj_close_minus_10"] = df_ticker['Adj Close'].shift(-10)
  df_ticker["adj_close_growth_10d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_10"] - 1
  df_ticker["adj_close_minus_11"] = df_ticker['Adj Close'].shift(-11)
  df_ticker["adj_close_growth_11d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_11"] - 1
  df_ticker["adj_close_minus_12"] = df_ticker['Adj Close'].shift(-12)
  df_ticker["adj_close_growth_12d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_12"] - 1
  df_ticker["adj_close_minus_13"] = df_ticker['Adj Close'].shift(-13)
  df_ticker["adj_close_growth_13d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_13"] - 1
  df_ticker["adj_close_minus_14"] = df_ticker['Adj Close'].shift(-14)
  df_ticker["adj_close_growth_14d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_14"] - 1
  df_ticker["adj_close_minus_15"] = df_ticker['Adj Close'].shift(-15)
  df_ticker["adj_close_growth_15d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_15"] - 1
  df_ticker["adj_close_minus_16"] = df_ticker['Adj Close'].shift(-16)
  df_ticker["adj_close_growth_16d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_16"] - 1
  #df_ticker = df_ticker.drop("adj_close_minus_1", axis=1)
  df_ticker["adj_close_minus_17"] = df_ticker['Adj Close'].shift(-17)
  df_ticker["adj_close_growth_17d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_17"] - 1
  df_ticker["adj_close_minus_18"] = df_ticker['Adj Close'].shift(-18)
  df_ticker["adj_close_growth_18d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_18"] - 1
  df_ticker["adj_close_minus_19"] = df_ticker['Adj Close'].shift(-19)
  df_ticker["adj_close_growth_19d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_19"] - 1
  df_ticker["adj_close_minus_20"] = df_ticker['Adj Close'].shift(-20)
  df_ticker["adj_close_growth_20d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_20"] - 1
  df_ticker["adj_close_minus_21"] = df_ticker['Adj Close'].shift(-21)
  df_ticker["adj_close_growth_21d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_21"] - 1
  df_ticker["adj_close_minus_22"] = df_ticker['Adj Close'].shift(-22)
  df_ticker["adj_close_growth_22d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_22"] - 1
  df_ticker["adj_close_minus_23"] = df_ticker['Adj Close'].shift(-23)
  df_ticker["adj_close_growth_23d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_23"] - 1
  df_ticker["adj_close_minus_24"] = df_ticker['Adj Close'].shift(-24)
  df_ticker["adj_close_growth_24d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_24"] - 1
  df_ticker["adj_close_minus_25"] = df_ticker['Adj Close'].shift(-25)
  df_ticker["adj_close_growth_25d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_25"] - 1
  df_ticker["adj_close_minus_26"] = df_ticker['Adj Close'].shift(-26)
  df_ticker["adj_close_growth_26d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_26"] - 1
  df_ticker["adj_close_minus_27"] = df_ticker['Adj Close'].shift(-27)
  df_ticker["adj_close_growth_27d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_27"] - 1
  df_ticker["adj_close_minus_28"] = df_ticker['Adj Close'].shift(-28)
  df_ticker["adj_close_growth_28d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_28"] - 1
  df_ticker["adj_close_minus_29"] = df_ticker['Adj Close'].shift(-29)
  df_ticker["adj_close_growth_29d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_14"] - 1
  df_ticker["adj_close_minus_30"] = df_ticker['Adj Close'].shift(-30)
  df_ticker["adj_close_growth_30d"] = df_ticker['Adj Close']/df_ticker["adj_close_minus_30"] - 1
  ticker_df.append(df_ticker)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

In [60]:
df_ipos_strat = pd.concat(ticker_df, axis=0, ignore_index=False)

In [61]:
df_ipos_strat

Open    High     Low   Close  Adj Close   Volume Symbol  \
Date                                                                    
2024-02-16  10.050  10.050  10.010  10.010     10.010    16700   IROH   
2024-02-20  10.020  10.020  10.020  10.020     10.020     5200   IROH   
2024-02-21  10.020  10.020  10.015  10.015     10.015    98600   IROH   
2024-02-22  10.020  10.020  10.020  10.020     10.020     5600   IROH   
2024-02-23  10.020  10.020  10.010  10.010     10.010    14800   IROH   
...            ...     ...     ...     ...        ...      ...    ...   
2024-05-01   0.710   0.728   0.661   0.678      0.678    55500   ROMA   
2024-05-02   0.693   0.721   0.658   0.711      0.711   202900   ROMA   
2024-05-03   0.750   0.777   0.670   0.719      0.719  1066800   ROMA   
2024-05-06   0.707   0.725   0.683   0.696      0.696   105300   ROMA   
2024-05-07   0.720   0.774   0.696   0.750      0.750   211000   ROMA   

            adj_close_minus_1  adj_close_growth_1d  adj_close_minus_2  ...  \
Date                                                                   ...   
2024-02-16             10.020            -0.000998             10.015  ...   
2024-02-20             10.015             0.000499             10.020  ...   
2024-02-21             10.020            -0.000499             10.010  ...   
2024-02-22             10.010             0.000999             10.020  ...   
2024-02-23             10.020            -0.000998             10.010  ...   
...                       ...                  ...                ...  ...   
2024-05-01              0.711            -0.046414              0.719  ...   
2024-05-02              0.719            -0.011127              0.696  ...   
2024-05-03              0.696             0.033046              0.750  ...   
2024-05-06              0.750            -0.072000                NaN  ...   
2024-05-07                NaN                  NaN                NaN  ...   

            adj_close_minus_26  adj_close_growth_26d  adj_close_minus_27  \
Date                                                                       
2024-02-16               9.980              0.003006               9.985   
2024-02-20               9.985              0.003505               9.980   
2024-02-21               9.980              0.003507               9.990   
2024-02-22               9.990              0.003003               9.990   
2024-02-23               9.990              0.002002              10.025   
...                        ...                   ...                 ...   
2024-05-01                 NaN                   NaN                 NaN   
2024-05-02                 NaN                   NaN                 NaN   
2024-05-03                 NaN                   NaN                 NaN   
2024-05-06                 NaN                   NaN                 NaN   
2024-05-07                 NaN                   NaN                 NaN   

            adj_close_growth_27d  adj_close_minus_28  adj_close_growth_28d  \
Date                                                                         
2024-02-16              0.002504               9.980              0.003006   
2024-02-20              0.004008               9.990              0.003003   
2024-02-21              0.002503               9.990              0.002503   
2024-02-22              0.003003              10.025             -0.000499   
2024-02-23             -0.001496              10.017             -0.000699   
...                          ...                 ...                   ...   
2024-05-01                   NaN                 NaN                   NaN   
2024-05-02                   NaN                 NaN                   NaN   
2024-05-03                   NaN                 NaN                   NaN   
2024-05-06                   NaN                 NaN                   NaN   
2024-05-07                   NaN                 NaN                   NaN   

            adj_close_minus_29  adj_close_gro

In [62]:
df_ipos_strat.describe()

Open          High           Low         Close     Adj Close  \
count  37770.000000  37770.000000  37770.000000  37770.000000  37770.000000   
mean      12.515059     13.139440     12.045827     12.485686     12.456813   
std       38.123264     56.866338     32.879304     40.746515     40.735089   
min        0.002000      0.002000      0.002000      0.002000      0.002000   
25%        1.890000      1.990000      1.780000      1.870000      1.870000   
50%        8.100000      8.362000      7.800000      8.060000      8.060000   
75%       12.140000     12.460000     11.660000     12.100000     12.090993   
max     3069.000000   7500.000000    875.000000   4318.000000   4318.000000   

             Volume  adj_close_minus_1  adj_close_growth_1d  \
count  3.777000e+04       37586.000000         37586.000000   
mean   6.385625e+05          12.336431             0.009279   
std    5.008113e+06          34.216165             0.226304   
min    0.000000e+00           0.002000            -0.853521   
25%    5.500000e+03           1.860000            -0.014510   
50%    4.940000e+04           8.055000             0.000000   
75%    2.213750e+05          12.085750             0.022820   
max    3.723413e+08         901.000000            33.875000   

       adj_close_minus_2  adj_close_growth_2d  ...  adj_close_minus_26  \
count       37402.000000         37402.000000  ...        32994.000000   
mean           12.310453             0.024988  ...           12.098707   
std            34.001322             1.471897  ...           33.619309   
min             0.002000            -0.856555  ...            0.002000   
25%             1.850000            -0.019655  ...            1.710000   
50%             8.060000             0.000000  ...            7.895000   
75%            12.084750             0.034483  ...           12.000000   
max           901.000000           278.000001  ...          901.000000   

       adj_close_growth_26d  adj_close_minus_27  adj_close_growth_27d  \
count          32994.000000        32811.000000          32811.000000   
mean               0.494959           12.095100              0.516772   
std               10.037492           33.647051             10.269421   
min               -0.933445            0.002000             -0.932351   
25%               -0.056048            1.710000             -0.058249   
50%                0.004864            7.881000              0.005160   
75%                0.226657           12.000000              0.231029   
max              365.499973          901.000000            365.499973   

       adj_close_minus_28  adj_close_growth_28d  adj_close_minus_29  \
count        32628.000000          32628.000000        32446.000000   
mean            12.091948              0.538639           12.088524   
std             33.681162             10.498767           33.719379   
min              0.002000             -0.949123            0.002000   
25%              1.700000             -0.059778            1.700000   
50%              7.875389              0.005366            7.860427   
75%             12.000000              0.237913           12.000000   
max            901.000000            365.499973          901.000000   

       adj_close_growth_29d  adj_close_minus_30  adj_close_growth_30d  
count          35194.000000        32264.000000          32264.000000  
mean               0.248963           12.071265              0.574628  
std                6.946977           33.542102             10.841248  
min               -0.911746            0.002000             -0.963233  
25%               -0.044246            1.690000             -0.062385  
50%                0.000988            7.850000              0.005958  
75%                0.132317           12.000000              0.250000  
max              365.499973          901.000000            365.499973  

[8 rows x 66 columns]

In [63]:
adj25percent, adj50percent, adj75percent = {}, {}, {}
desc = df_ipos_strat.describe().to_dict()
for i in desc.items():
  if i[0].startswith('adj_close_growth'):
    adj25percent[i[0]] = desc[i[0]]['25%']
    adj50percent[i[0]] = desc[i[0]]['50%']
    adj75percent[i[0]] = desc[i[0]]['75%']

In [64]:
def max_value_key(dict):
  max_value_key = max(dict, key=dict.get)
  max_value = dict[max_value_key]
  return (max_value_key, max_value)

In [68]:
(key25, max25value), (key50, max50value), (key75, max75value)  = max_value_key(adj25percent), max_value_key(adj50percent), max_value_key(adj75percent)
maxStrat = [key25, key50, key75][np.argmax([max25value, max50value, max75value])]
f'Best strategy is {maxStrat}'

'Best strategy is adj_close_growth_30d'

### Question 3: Is Growth Concentrated in the Largest Stocks?

**Get the share of days (percentage as int) when Large Stocks outperform (growth_7d - growth over 7 periods back) the Largest stocks?**


Reuse [Code Snippet 5] to obtain OHLCV stats for 33 stocks 
for 10 full years of data (2014-01-01 to 2023-12-31). You'll need to download slightly more data (7 periods before 2014-01-01 to calculate the growth_7d for the first 6 days correctly):

`US_STOCKS = ['MSFT', 'AAPL', 'GOOG', 'NVDA', 'AMZN', 'META', 'BRK-B', 'LLY', 'AVGO','V', 'JPM']`

`EU_STOCKS = ['NVO','MC.PA', 'ASML', 'RMS.PA', 'OR.PA', 'SAP', 'ACN', 'TTE', 'SIE.DE','IDEXY','CDI.PA']`

`INDIA_STOCKS = ['RELIANCE.NS','TCS.NS','HDB','BHARTIARTL.NS','IBN','SBIN.NS','LICI.NS','INFY','ITC.NS','HINDUNILVR.NS','LT.NS']`

`LARGEST_STOCKS = US_STOCKS + EU_STOCKS + INDIA_STOCKS`
<br/>

Now let's add the top 12-22 stocks (as of end-April 2024):
<br/>

`NEW_US = ['TSLA','WMT','XOM','UNH','MA','PG','JNJ','MRK','HD','COST','ORCL']`

`NEW_EU = ['PRX.AS','CDI.PA','AIR.PA','SU.PA','ETN','SNY','BUD','DTE.DE','ALV.DE','MDT','AI.PA','EL.PA']`

`NEW_INDIA = ['BAJFINANCE.NS','MARUTI.NS','HCLTECH.NS','TATAMOTORS.NS','SUNPHARMA.NS','ONGC.NS','ADANIENT.NS','ADANIENT.NS','NTPC.NS','KOTAKBANK.NS','TITAN.NS']`

`LARGE_STOCKS = NEW_EU + NEW_US + NEW_INDIA`

You should be able to obtain stats for 33 LARGEST STOCKS and 32 LARGE STOCKS (from the actual stats on Yahoo Finance)

Calculate  `growth_7d` for every stock and every day.
Get the average daily `growth_7d` for the LARGEST_STOCKS group vs. the LARGE_STOCKS group.

For example, for the first of data you should have:
| Date   |      ticker_category      |  growth_7d |
|----------|:-------------:|------:|
| 2014-01-01 |  LARGE | 1.011684 |
| 2014-01-01 |   LARGEST   |   1.011797 |

On that day, the LARGEST group was growing faster than LARGE one (new stocks).

Calculate the number of days when the LARGE GROUP (new smaller stocks) outperforms the LARGEST GROUP, divide it by the total number of trading days (which should be 2595 days), and convert it to a percentage (closest INTEGER value). For example, if you find that 1700 out of 2595 days meet this condition, it means that 1700/2595 = 0.655, or approximately 66% of days, the LARGE stocks were growing faster than the LARGEST ones. This suggests that you should consider extending your dataset with more stocks to seek higher growth.

HINT: you can use pandas.pivot_table() to "flatten" the table (LARGE and LARGEST growth_7d as columns)